From a9c4e4c9fc11cf1bd17d08e166405f7ab355a9f3 Mon Sep 17 00:00:00 2001
From: Herbert Xu <herbert@gondor.apana.org.au>
Date: Sat, 19 May 2018 02:39:46 +0800
Subject: [PATCH 09/53] memalloc: Add growstackto helper

This patch adds the growstackto helper which repeatedly calls
growstackblock until the requested size is reached.

Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
---
 src/exec.c     |  4 +---
 src/memalloc.c | 20 +++++++++-----------
 src/memalloc.h |  1 +
 src/parser.c   |  4 +---
 4 files changed, 12 insertions(+), 17 deletions(-)

diff --git a/src/exec.c b/src/exec.c
index d7ced35..c98f14c 100644
--- a/src/exec.c
+++ b/src/exec.c
@@ -195,9 +195,7 @@ padvance(const char **path, const char *name)
 	start = *path;
 	for (p = start ; *p && *p != ':' && *p != '%' ; p++);
 	len = p - start + strlen(name) + 2;	/* "2" is for '/' and '\0' */
-	while (stackblocksize() < len)
-		growstackblock();
-	q = stackblock();
+	q = growstackto(len);
 	if (p != start) {
 		memcpy(q, start, p - start);
 		q += p - start;
diff --git a/src/memalloc.c b/src/memalloc.c
index d8e4413..9d1de74 100644
--- a/src/memalloc.c
+++ b/src/memalloc.c
@@ -265,6 +265,14 @@ growstackstr(void)
 	return stackblock() + len;
 }
 
+char *growstackto(size_t len)
+{
+	while (stackblocksize() < len)
+		growstackblock();
+
+	return stackblock();
+}
+
 /*
  * Called from CHECKSTRSPACE.
  */
@@ -273,18 +281,8 @@ char *
 makestrspace(size_t newlen, char *p)
 {
 	size_t len = p - stacknxt;
-	size_t size;
 
-	for (;;) {
-		size_t nleft;
-
-		size = stackblocksize();
-		nleft = size - len;
-		if (nleft >= newlen)
-			break;
-		growstackblock();
-	}
-	return stackblock() + len;
+	return growstackto(len + newlen) + len;
 }
 
 char *
diff --git a/src/memalloc.h b/src/memalloc.h
index 4b5be46..b348d9c 100644
--- a/src/memalloc.h
+++ b/src/memalloc.h
@@ -57,6 +57,7 @@ void setstackmark(struct stackmark *);
 void popstackmark(struct stackmark *);
 void growstackblock(void);
 void *growstackstr(void);
+char *growstackto(size_t len);
 char *makestrspace(size_t, char *);
 char *stnputs(const char *, size_t, char *);
 char *stputs(const char *, char *);
diff --git a/src/parser.c b/src/parser.c
index 809c6a8..3de977c 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -1460,9 +1460,7 @@ done:
 	/* Ignore any pushed back tokens left from the backquote parsing. */
 	if (oldstyle)
 		tokpushback = 0;
-	while (stackblocksize() <= savelen)
-		growstackblock();
-	STARTSTACKSTR(out);
+	out = growstackto(savelen + 1);
 	if (str) {
 		memcpy(out, str, savelen);
 		STADJUST(savelen, out);
-- 
2.25.0

